### PR TITLE
simplify RDS and EDS request process

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
This makes EDS and RDS processing consistent.

### CODE REVIEW

File: pilot/pkg/proxy/envoy/v2/ads.go [hunk 1]
```diff
@@ -546,11 +551,14 @@ func (s *DiscoveryServer) StreamAggregatedResources(stream ads.AggregatedDiscove
 
 					// Already got a list of endpoints to watch and it is the same as the request, this is an ack
 					if reflect.DeepEqual(con.Clusters, clusters) {
-						if discReq.ErrorDetail == nil && discReq.ResponseNonce != "" {
+						adsLog.Debugf("ADS:EDS: ACK %s %s (%s) %s %s", peerAddr, con.ConID, con.modelNode, discReq.VersionInfo, discReq.ResponseNonce)
```
Comment (costinm) [inline pilot/pkg/proxy/envoy/v2/ads.go]: This will cause very verbose messages

File: pilot/pkg/proxy/envoy/v2/ads.go [hunk 2]
```diff
@@ -546,11 +551,14 @@ func (s *DiscoveryServer) StreamAggregatedResources(stream ads.AggregatedDiscove
...
+						if discReq.ResponseNonce != "" {
```
Comment (costinm) [inline pilot/pkg/proxy/envoy/v2/ads.go]: What happened with ErrorDetail==nil ? Nacks are not acks

File: pilot/pkg/proxy/envoy/v2/ads.go [hunk 3]
```diff
@@ -522,18 +527,18 @@ func (s *DiscoveryServer) StreamAggregatedResources(stream ads.AggregatedDiscove
...
+					continue
 				}
-
+				clusters := discReq.GetResourceNames()
 				if clusters == nil && discReq.ResponseNonce != "" {
 					// There is no requirement that ACK includes clusters. The test doesn't.
-					if discReq.ErrorDetail == nil && discReq.ResponseNonce != "" {
-						con.EndpointNonceAcked = discReq.ResponseNonce
-					}
+					con.mu.Lock()
```
Comment (costinm) [inline pilot/pkg/proxy/envoy/v2/ads.go]: Same - AFAIK NonceAcked is not the same thing with nonce nacked. It's a semantic change

File: pilot/pkg/proxy/envoy/v2/ads.go [hunk 4]
```diff
@@ -483,34 +483,39 @@ func (s *DiscoveryServer) StreamAggregatedResources(stream ads.AggregatedDiscove
...
-					// routes and con.Routes are all empty, this is not an ack and will do nothing.
-					if len(routes) == 0 && len(con.Routes) == 0 {
-						continue
-					}
-
+				if len(routes) == len(con.Routes) {
 					sort.Strings(routes)
 					sort.Strings(con.Routes)
 
-					if reflect.DeepEqual(con.Routes, routes) || len(routes) == 0 {
```
Comment (costinm) [inline pilot/pkg/proxy/envoy/v2/ads.go]: Why did you remove the prometheus tracking of nacks ? That was pretty important.

**Comment by costinm** [comment]
Can you rollback this - pretty sure it breaks at least few things, and not clear what bugs it fixes.

File: pilot/pkg/proxy/envoy/v2/ads.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pilot/pkg/proxy/envoy/v2/ads.go]: I do not change this semantic , L530-L534 already guaranteed NACK is returned. 
If run into here, it must be ACK

File: pilot/pkg/proxy/envoy/v2/ads.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/pkg/proxy/envoy/v2/ads.go]: L530 and L537 checks `if discReq.ErrorDetail == nil && discReq.ResponseNonce != ""` this already

File: pilot/pkg/proxy/envoy/v2/ads.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/pkg/proxy/envoy/v2/ads.go]: I think you mean What happened with ErrorDetail!=nil ? 

It is checked at first in L530 if ErrorDetail!=nil continue skip the following.

**Comment by hzxuzhonghu** [comment]
This is absolutely not a bug fix.   Actually I can not find what it breaks according to your comments above.

**Comment by hzxuzhonghu** [comment]
OK. I will not do similar things close to release. Maybe we should refer k8s, freeze our code vefore release. Only important bug fix can be merged.
